### DISCUSSION TITLE
Requirements for function passed to to `DataArrayGroupBy.map()`

### DISCUSSION THREAD

**Discussion by jlumpe** [pydata/xarray#5864] [Q&A]
The user documentation doesn't really give much information on what kind of functions you can use with `map()`, it seems like it only covers numpy ufuncs or compositions of them. The example of using a custom function is:

```python3
def standardize(x):
     return (x - x.mean()) / x.std()
```

which will return a `DataArray` for `DataArray` input. The documentation doesn't explicitly state that this is required, but when I try to use it with a function that returns a scalar I get:

```

~/opt/anaconda3/envs/gambit/lib/python3.9/site-packages/xarray/core/groupby.py in map(self, func, shortcut, args, **kwargs)
    792         grouped = self._iter_grouped_shortcut() if shortcut else self._iter_grouped()
    793         applied = (maybe_wrap_array(arr, func(arr, *args, **kwargs)) for arr in grouped)
--> 794         return self._combine(applied, shortcut=shortcut)
    795 
    796     def apply(self, func, shortcut=False, args=(), **kwargs):

~/opt/anaconda3/envs/gambit/lib/python3.9/site-packages/xarray/core/groupby.py in _combine(self, applied, shortcut)
    812         """Recombine the applied objects like the original."""
    813         applied_example, applied = peek_at(applied)
--> 814         coord, dim, positions = self._infer_concat_args(applied_example)
    815         if shortcut:
    816             combined = self._concat_shortcut(applied, dim, positions)

~/opt/anaconda3/envs/gambit/lib/python3.9/site-packages/xarray/core/groupby.py in _infer_concat_args(self, applied_example)
    460 
    461     def _infer_concat_args(self, applied_example):
--> 462         if self._group_dim in applied_example.dims:
    463             coord = self._group
    464             positions = self._group_indices

AttributeError: 'int' object has no attribute 'dims'
```

Wrapping the return value in a zero-dimensional `DataArray` fixes the issue. Is this the expected behavior?

**Comment by mathause** [comment]
Do you remember the function you used? E.g. the following works 

```python
air = xr.tutorial.load_dataset("air_temperature")
air.map(lambda x: 1)
```

**Reply by keewis** [reply]
I think the call is something like `air.groupby("time.month").map(lambda x: 1)` which does indeed raise the posted error. This makes sense to me because in addition to applying the function to the groups it tries to concatenate the result.

**Comment by jlumpe** [comment]
@keewis is correct, this is using the result of `DataArray.groupby()`. My expectation is that any return value that is not an instance of `DataArray` should be converted to one (probably just by calling `DataArray(value)`), similar to how Pandas would convert the return value of the function passed to `DataFrame.apply()`.
